File: udemy/ds-and-a/DSAndA/LinkedLists/DoublyEndedList.cs
using DSAndA.LinkedLists;

namespace DSAndA
{
    public class DoublyEndedList
    {
        private Node<int> head;
        private Node<int> tail;

        public void InsertAtHead(int data)
        {
            Node<int> newNode = new Node<int>(data);

            newNode.NextNode = this.head;
            this.head = newNode;

            if (this.tail == null)
                this.tail = newNode;
        }

        public void InsertAtTail(int data)
        {
            Node<int> newNode = new Node<int>(data);

            if (this.head == null)
            {
                this.head = newNode;
                this.tail = newNode;

                return;
            }

            this.tail.NextNode = newNode;
            this.tail = newNode;
        }

        public override string ToString()
        {
            string nodes = string.Empty;

            Node<int> current = this.head;

            while (current != null)
            {
                nodes += $"{current.ToString()},";

                current = current.NextNode;
            }

            return nodes;
        }
    }
}

File: udemy/ds-and-a/javascript-solutions/src/LinkedLists/DoublyEndedList.js
const Node = require("./Node");

class DoublyEndedList {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    // O(1)
    insertAtTail(data) {
        const newNode = new Node(data);

        if (!this.tail) {
            this.head = newNode;
            this.tail = newNode;

            return;
        }

        this.tail.nextNode = newNode;
        this.tail = newNode;
    }

    toString() {
        let result = [];
        let current = this.head;

        while (current != null) {
            result.push(current.data);

            current = current.nextNode;
        }

        return result.join(', ');
    }
}

module.exports = DoublyEndedList;